[rpc] Add getnodeaddresses RPC command

chris-belcher opened the pull request:
Implements issue https://github.com/bitcoin/bitcoin/issues/9463

New getnodeaddresses call gives access via RPC to the peers known by the node. It may be useful for bitcoin wallets to broadcast their transactions over tor for improved privacy without using the centralized DNS seeds. getnodeaddresses is very similar to the getaddr p2p method.

Please advise me on the best approach for writing an automated test. By my reading the getaddr p2p method also isn't really tested.

chris-belcher committed:
[rpc] Add getnodeaddresses RPC command

New getnodeaddresses call gives access via RPC to the peers known by
the node. It may be useful for bitcoin wallets to broadcast their
transactions over tor for improved privacy without using the
centralized DNS seeds. getnodeaddresses is very similar to the getaddr
p2p method.

Tests the new rpc call by feeding IP address to a test node via the p2p
protocol, then obtaining someone of those addresses with
getnodeaddresses and checking that they are a subset.

promag commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,58 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+    UniValue ret(UniValue::VARR);
+
+    int address_return_count = std::min(count, (int)vAddr.size());
+    for(int i = 0; i < address_return_count; ++i) {
```
Nit, space after `for`.

promag commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,58 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+            "getaddress ( count )\n"
+            "\nReturn known addresses which can potentially be used to find new nodes in the network\n"
+            "\nArguments:\n"
+            "1. \"count\"    (numeric, optional) How many addresses to return, if available (default = 1)\n"
```
Why not return all?

chris-belcher replied:
Returning all would typically result in a massive json array with thousands of entries, about 2500 or so depending. Normally applications only need to connect to a few nodes, they don't need anywhere near that many.

promag reviewed (commented):
Why not extend `getpeerinfo`?

promag commented:
> Please advise me on the best approach for writing an automated test

At least it can test arguments (valid and invalid) and result format and fields (their presence and types).

chris-belcher commented:
Renamed to `getnodeaddresses` and fixed nits. Looks like something went wrong with the rebase and now bluematt's commits are involved too.

> At least it can test arguments (valid and invalid) and result format and fields (their presence and types).

I think in the test suite the node won't know about any other addresses, so this RPC will return an empty json object, so it can't be tested. What would be the best way around this? Maybe to populate the `vAddr` somehow.

chris-belcher commented:
Fixed git weirdness (thanks to viasil).

A reviewer asked why does the `count` argument exist, why not return all the addresses? My reasoning is that this would result in a massive json array with thousands of entries, about 2500 or so depending. Normally applications only need to connect to a few nodes, they don't need anywhere near that many.

promag commented:
Please rename PR.

MasterGrad17 commented:
Does getnodeaddresses command return a list of reachable nodes in the network known by a node? like the result of getaddr message?

chris-belcher commented:
@MasterGrad17 the result is almost exactly the same as from the p2p getaddr method. The addresses come from gossiping between nodes, so the IP address _may_ be reachable but that isn't guaranteed.

sipa commented:
@MasterGrad17 You're commenting on the request to get it merged into Bitcoin Core, so obviously no.

jonasschnelli commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,57 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+            "    \"time\": ttt,                        (numeric) Address timestamp in seconds since epoch (Jan 1 1970 GMT)\n"
+            "    \"services\": n,                      (numeric) The services offered\n"
+            "    \"servicesHex\": \"xxxxxxxxxxxxxxxx\",  (string) The hex string of services offered\n"
+            "    \"addr\": \"host\",                     (string) The IP address of the peer\n"
```
nit: could also be a tor address

jonasschnelli commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,57 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+        CAddress addr = vAddr[i];
+        obj.pushKV("time", (int)addr.nTime);
+        obj.pushKV("services", addr.nServices);
+        obj.pushKV("servicesHex", strprintf("%016x", addr.nServices));
```
would be handy to have a string representation here. Like (`"NODE_NETWORK | NODE_NETWORK_LIMITED | NODE_BLOOM"`, etc.). But can be done later.

jonasschnelli reviewed (changes requested):
Almost utACK c83ccf76b0df2aceedddc23c8e081e877e5a878d.
Please rename command.
Maybe `listpeeraddresses()`

sipa commented:
@jonasschnelli `listpeeraddresses` sounds like it's somehow exhaustive, or at least a complete specific subset of addresses. It's just giving a bunch of random ones; `getpeeraddresses` sounds clearer to me. Am i missing something?

jonasschnelli commented:
> listpeeraddresses sounds like it's somehow exhaustive, or at least a complete specific subset of addresses. It's just giving a bunch of random ones; getpeeraddresses sounds clearer to me. Am I missing something?

Good point.

chris-belcher commented:
Fixed nit

jnewbery reviewed (commented):
A few nits inline.

You should be able to test this in the functional test framework as follows:

- start a node
- add a P2PConnection to the node
- send a msg_addr to the node with some addresses
- call the new `getnodeaddresses` RPC
- assert that the node addresses returned are the same as those sent in the p2p ADDR message.

Take a look at example_test.py for some pointers on how to write test cases. I think it makes sense to add this new test case to the rpc_net.py script.

Feel free to reach me on irc (jnewbery) if you need any pointers.

laanwj commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,57 @@ UniValue setnetworkactive(const JSONRPCRequest& request)
+            "[\n"
+            "  {\n"
+            "    \"time\": ttt,                        (numeric) Address timestamp in seconds since epoch (Jan 1 1970 GMT)\n"
+            "    \"services\": n,                      (numeric) The services offered\n"
```
Depends on what 'the user' is. Programmatic RPC users will want the int, not have to do an additional step of parsing a hex string.

jnewbery replied:
> be mindful that manual command-line users are not the only clients of the RPC interface

Yes good point. In general though, I think we should avoid RPCs returning the same data in multiple formats.

chris-belcher commented:
Fixed nits and created a test in `rpc_net.py`.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -624,6 +624,55 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
     return g_connman->GetNetworkActive();
 }
 
+UniValue getnodeaddresses(const JSONRPCRequest& request)
```
nit: should be `static`

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -16,6 +17,8 @@
     p2p_port,
     wait_until,
 )
+from test_framework.mininode import P2PInterface, network_thread_start
```
nit: sort imports (in PEP-8 ordering)

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -96,5 +100,29 @@ def _test_getpeerinfo(self):
+        for a in imported_addrs:
+            addr = CAddress()
+            addr.time = now
+            addr.services = NODE_NETWORK | NODE_WITNESS
```
Should be `nServices`

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -96,5 +100,29 @@ def _test_getpeerinfo(self):
+        self.nodes[0].p2p.wait_for_verack()
+
+        # send some addresses to the node via the p2p message addr
+        imported_addrs = set(["123.123." + str(i//255) + "." + str(i%255) for i in range(1000)])
```
nit: move this into the range loop below and use string formatters rather than concatenation:

```python
        # send some addresses to the node via the p2p message addr
        now = int(time.time())
        msg = msg_addr()
        imported_addrs = []
        for i in range(1000):
            a = "123.123.{}.{}".format(i // 255, i % 256)
            imported_addrs.append(a)
            addr = CAddress()
            ...
```

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -96,5 +100,29 @@ def _test_getpeerinfo(self):
+        NODE_ADDRESSES_REQUEST_COUNT = 10
+        node_addresses = self.nodes[0].getnodeaddresses(NODE_ADDRESSES_REQUEST_COUNT)
+        assert len(node_addresses) == NODE_ADDRESSES_REQUEST_COUNT
+        assert set([addr["address"] for addr in node_addresses]).issubset(imported_addrs)
```
nit: drop the set comparisons, loop over the return items and test all fields, eg:

```python
        for a in node_addresses:
            assert a["address"] in imported_addrs
            assert_equal(a["services"], NODE_NETWORK | NODE_WITNESS)
            assert_equal(a["time"], now)
            assert_equal(a["port"], 8333)
```

jnewbery reviewed (commented):
New test is great. Thanks for adding it!

A few nits inline. Feel free to take or leave them.

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -6,6 +6,7 @@
 
 Tests correspond to code in rpc/net.cpp.
 """
+import time
```
nit: You're only using `time` to set the time field in the ADDR message. As long as the `CAddress`'s time is greater than             100000000, it'll be accepted. You can avoid this import by just setting time to `100000000 + i` or similar

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -96,5 +98,35 @@ def _test_getpeerinfo(self):
+        msg = msg_addr()
+        imported_addrs = []
+        for i in range(1000):
+            a = "123.123.{}.{}".format(i // 255, i % 256)
```
Note that you'll only ever be able to get a maximum of 64 entries into the new addresses buckets by doing this. All of these addresses are in the same /16 address range, which means that addrman will place them into the same new bucket, and start evicting addresses once there are 64 entries (see the comment in addrman.h).

You may wish to simplify this to just send 256 addresses:

```python
        for i in range(256):
            a = "123.123.123.{}".format(i)
```

jnewbery reviewed (commented):
One more comment inline, which answers the question you had on #bitcoin-core-dev IRC earlier.

> bitcoind modifies the timestamp so it's complicated to check whether the time was equal.

bitcoind modifies the timestamp if it falls outside these bounds:

```
            if (addr.nTime <= 100000000 || addr.nTime > nNow + 10 * 60)
                addr.nTime = nNow - 5 * 24 * 60 * 60;
```

but getting a valid time in that range is a bit difficult. if the time is greater than 30 days old, then the address will be marked as `IsTerrible` and won't be returned in your RPC call. You could set the node's mock time and then test that the returned time is that + (5 * 24 * 60 * 60).

jnewbery commented:
@jonasschnelli - can you remove your 'changes requested' status?

jonasschnelli reviewed (approved):
utACK f10e38063048294afc6bd45c3a51687cdbdd96a3

(sorry for the delay)

jnewbery commented:
Needs rebase. The `rpc_net.py` functional test is failing.

ajtowns commented:
The `rpc_net.py` test looks like it just needs the references to `network_thread_start` removed, in line with #13517

chris-belcher commented:
Fixed the broken test

ajtowns commented:
utACK 316201485933b4fcdf8d4005d823effdfc4b913b

jnewbery commented on test/functional/rpc_net.py:
```diff
@@ -96,5 +98,34 @@ def _test_getpeerinfo(self):
 
+    def _test_getnodeaddresses(self):
+        self.nodes[0].add_p2p_connection(P2PInterface())
+        self.nodes[0].p2p.wait_for_verack()
```
This is no longer required (`add_p2p_connection()` calls `wait_for_verack()` now)

jnewbery reviewed (commented):
Looks great. tACK 316201485933b4fcdf8d4005d823effdfc4b913b.

One nit inline. I also think the two commits can be squashed when you rebase this.

chris-belcher commented:
Rebased.

The test has failed but works when I run it locally.

promag commented:
Unrelated build error:
```
...
Get:17 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [389 kB]
Get:18 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [2807 B]
Fetched 25.9 MB in 8min 28s (50.9 kB/s)
Reading package lists...
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```
Restarted job.

promag commented on src/rpc/net.cpp:
```diff
@@ -626,6 +626,55 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
+
+    int count = 1;
+    if (!request.params[0].isNull()) {
+        count = request.params[0].get_int();
```
Should validate `count > 0`? Also, should have a maximum value?

promag replied:
nit, could have a test for for invalid argument.

promag commented on test/functional/rpc_net.py:
```diff
@@ -101,5 +104,33 @@ def _test_getpeerinfo(self):
+
+        # obtain addresses via rpc call and check they were ones sent in before
+        NODE_ADDRESSES_REQUEST_COUNT = 10
+        node_addresses = self.nodes[0].getnodeaddresses(NODE_ADDRESSES_REQUEST_COUNT)
```
Use named arguments instead of long and verbose variables? Then ditch `NODE_ADDRESSES_REQUEST_COUNT`.

promag replied:
Could also test that the result is limited by the existing addressed, not by the requested count.

chris-belcher replied:
Unless I'm missing something, that verbose variable is still needed for the assert_equal check in the line after. It seems a better way than having magic numbers. I've reduced the length of the variable name.

promag commented on src/rpc/net.cpp:
```diff
@@ -626,6 +626,55 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
+    int address_return_count = std::min(count, (int)vAddr.size());
+    for (int i = 0; i < address_return_count; ++i) {
+        UniValue obj(UniValue::VOBJ);
+        CAddress addr = vAddr[i];
```
`const CAddress&`?

promag commented on test/functional/rpc_net.py:
```diff
@@ -101,5 +104,33 @@ def _test_getpeerinfo(self):
+        node_addresses = self.nodes[0].getnodeaddresses(NODE_ADDRESSES_REQUEST_COUNT)
+        assert_equal(len(node_addresses), NODE_ADDRESSES_REQUEST_COUNT)
+        for a in node_addresses:
+            # the timestamps are usually offset by a few hours, so only check existance
```
But could check that value is greater than some value?

promag commented on src/rpc/net.cpp:
```diff
@@ -626,6 +626,55 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
+    std::vector<CAddress> vAddr = g_connman->GetAddresses();
+    UniValue ret(UniValue::VARR);
+
+    int address_return_count = std::min(count, (int)vAddr.size());
```
Why the cast? Could inline std::min below.

promag reviewed (commented):
Care to add a release note by creating the file doc/release-notes-13152.md?

chris-belcher commented:
Thanks for the review @promag. I have created the release notes entry in the style of the release-notes-0.15.0 file.

promag commented on src/rpc/net.cpp:
```diff
@@ -626,6 +626,57 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Address count out of range");
+    }
+    // returns a shuffled list of CAddress
+    std::vector<CAddress> vAddr = g_connman->GetAddresses();
```
Makes sense to pass the max here?

chris-belcher replied:
Can you rephrase this, I didn't understand? What would be done with the max? The count parameter is already a bit like a max?

promag replied:
I mean, GetAddresses could return count addresses instead.

chris-belcher replied:
How? GetAddresses doesn't accept a parameter

promag replied:
Right, I'm asking if it makes sense to add that parameter.

chris-belcher replied:
Ah I see! I think it doesn't make sense, because GetAddresses is used in a few other places and that change would go against the principle of each PR only doing one thing.

promag replied:
Sounds reasonable.

practicalswift commented on test/functional/rpc_net.py:
```diff
@@ -101,5 +105,42 @@ def _test_getpeerinfo(self):
+        for a in node_addresses:
+            # the timestamps are usually offset by a few hours, so only check existence
+            assert "time" in a
+            assert_greater_than(a["time"], 1527811200) #1st June 2018
```
A very small nit, but to please PEP-8 consider fixing this:

```
./test/functional/rpc_net.py:132:56: E262 inline comment should start with '# '
```

chris-belcher replied:
Done

ken2812221 commented on src/rpc/net.cpp:
```diff
@@ -626,6 +626,57 @@ static UniValue setnetworkactive(const JSONRPCRequest& request)
+        UniValue obj(UniValue::VOBJ);
+        const CAddress& addr = vAddr[i];
+        obj.pushKV("time", (int)addr.nTime);
+        obj.pushKV("services", addr.nServices);
```
Appveyor error:
```
c:\projects\bitcoin\src\rpc\net.cpp(672): error C2668: 'UniValue::pushKV': ambiguous call to overloaded function
```
Maybe make it an explicit type?
 ```cpp
obj.pushKV("services", (uint64_t)addr.nServices);

promag commented:
Looks good to me, utACK 8b96ebe.

jnewbery reviewed (commented):
Sorry to nit at this late stage. Thanks so much for sticking with this. I know it's taken a long time!

One tiny code-style change and I think this is ready for merge.

promag commented on test/functional/rpc_net.py:
```diff
@@ -101,5 +105,42 @@ def _test_getpeerinfo(self):
+        assert_equal(len(node_addresses), REQUEST_COUNT)
+        for a in node_addresses:
+            # the timestamps are usually offset by a few hours, so only check existence
+            assert "time" in a
```
nit, can be removed.

chris-belcher commented:
Fixed those nits

ajtowns commented:
utACK a2eb6f540538d32725aecf678301a96247db6fd9

promag commented:
utACK a2eb6f5. Please fix PR description: "New getaddress call gives ..."
